src/index.ts

```diff
@@ -1568,44 +1565,61 @@ export class BigQuery extends common.Service {
    */
   query(
     query: string | Query,
-    optionsOrCallback?:
-      | QueryOptions
-      | SimpleQueryRowsCallback
-      | QueryRowsCallback,
-    cb?: SimpleQueryRowsCallback | QueryRowsCallback
-  ): void | Promise<SimpleQueryRowsResponse> | Promise<QueryRowsResponse> {
-    let options =
+    optionsOrCallback?: QueryOptions | QueryRowsCallback,
+    cb?: QueryRowsCallback
+  ): void | Promise<QueryRowsResponse> {
+    const options =
       typeof optionsOrCallback === 'object' ? optionsOrCallback : {};
     const callback =
       typeof optionsOrCallback === 'function' ? optionsOrCallback : cb;
-    this.createQueryJob(query, (err, job, resp) => {
-      if (err) {
-        (callback as SimpleQueryRowsCallback)(err, null, resp);
-        return;
-      }
-      if (typeof query === 'object' && query.dryRun) {
-        (callback as SimpleQueryRowsCallback)(null, [], resp);
-        return;
-      }
-      // The Job is important for the `queryAsStream_` method, so a new query
-      // isn't created each time results are polled for.
-      options = extend({job}, options);
-      job!.getQueryResults(options, callback as QueryRowsCallback);
-    });
+
+    query = typeof query === 'string' ? {query} : query;
+
+    if (query.table) {
+      query.table.getRows(options, callback!);
+      return;
+    }
+
+    this.runQuery_(query, options).then(
+      ([rows, nextQuery, resp]) => callback!(null, rows, nextQuery, resp),
+      callback!
+    );
   }
 
   /**
-   * This method will be called by `createQueryStream()`. It is required to
-   * properly set the `autoPaginate` option value.
-   *
    * @private
    */
-  queryAsStream_(query: Query, callback?: SimpleQueryRowsCallback) {
-    if (query.job) {
-      query.job.getQueryResults(query, callback as QueryRowsCallback);
-      return;
+  async runQuery_(query: Query, options: QueryOptions): Promise<RowsResponse> {
```

Comment:
@shollyman a second pair of eyes on this method would be great, basically it does the following

1. creates a query job
2. polls `getQueryResults` until `jobComplete` is `true` (should we set `maxResults` to something like 0 or 1?)
3. upon completion, if rows are found then we call `listTableData` (aka `table.getRows()`) via returned `destinationTable`

Reply:
> should we set maxResults to something like 0 or 1?

We set it to 0 in Python.

https://github.com/googleapis/google-cloud-python/blob/387b8e5de88d6af58b45e513426788c36a733fab/bigquery/google/cloud/bigquery/client.py#L857

Reply:
Excellent, thanks @tswast!